cpp/src/arrow/compute/exec/hash_join_node.cc

```diff
@@ -454,6 +456,172 @@ Status HashJoinSchema::CollectFilterColumns(std::vector<FieldRef>& left_filter,
   return Status::OK();
 }
 
+class HashJoinNode;
+
+struct BloomFilterPushdownContext {
+  using BuildFinishedCallback = std::function<Status(size_t, AccumulationQueue)>;
+  using FiltersReceivedCallback = std::function<Status()>;
+  using FilterFinishedCallback = std::function<Status(size_t, AccumulationQueue)>;
+  void Init(HashJoinNode* owner, size_t num_threads, TaskScheduler* scheduler,
+            FiltersReceivedCallback on_bloom_filters_received, bool disable_bloom_filter,
+            bool use_sync_execution);
+
+  Status StartProducing();
+
+  void ExpectBloomFilter() { eval_.num_expected_bloom_filters_ += 1; }
+
+  Status BuildBloomFilter(size_t thread_index, AccumulationQueue batches,
+                          BuildFinishedCallback on_finished);
+
+  Status PushBloomFilter();
+
+  Status ReceiveBloomFilter(std::unique_ptr<BlockedBloomFilter> filter,
+                            std::vector<int> column_map) {
```

Comment:
I think there would be some value here in comments explaining the general strategy of what we are trying to do (or maybe, as has also been discussed, we just need more general design documents).  Maybe just some short comments on each of these methods and a paragraph on the struct itself.

Reply:
Added a comment.